# Real world data - TAWES

## Tawes weather data example

Tawes weather data for kapfenberg from 1997 to 2018 is in `data/tawes/Messstationen Tagesdaten v2 Datensatz_19970101_20241001_kapfenberg.csv`.
CSV (Comma Separated Value) files can be read with pandas (amongst many other file formats).

In [1]:
import pandas
df = pandas.read_csv('../data/tawes/Messstationen Tagesdaten v2 Datensatz_19970101_20241001_kapfenberg.csv')
df.head()

time  station  cglo_j  rr  tl_mittel  vv_mittel  \
0  1997-01-01T00:00+00:00    13305     NaN NaN        NaN        NaN   
1  1997-01-02T00:00+00:00    13305     NaN NaN        NaN        NaN   
2  1997-01-03T00:00+00:00    13305     NaN NaN        NaN        NaN   
3  1997-01-04T00:00+00:00    13305     NaN NaN        NaN        NaN   
4  1997-01-05T00:00+00:00    13305     NaN NaN        NaN        NaN   

   p_mittel  tlmin  tlmax  
0       NaN    NaN    NaN  
1       NaN    NaN    NaN  
2       NaN    NaN    NaN  
3       NaN    NaN    NaN  
4       NaN    NaN    NaN

rr is sum percipitation for the whole day, cglo_j is global irradiance, tl_mittel is average air temperature, tlmin and tlmax are the respective temperature extrema on these days, vv_mittel is average wind speed, p_mittel is average air pressure.
However, there are no values. NaN is short for Not A Number. And if there is no value in this row (that is otherwise numeric in nature) it simple fills it with this NaN indicator.
The first row in the csv file looks like this:
`1997-01-01T00:00+00:00,13305,,,,,`
because the weather station started being operational only later that year, and only been operational until 2018.

So how to get rid of those NaNs?


In [2]:
df.dropna().head()

time  station  cglo_j    rr  tl_mittel  vv_mittel  \
1826  2002-01-01T00:00+00:00    13305   322.0  -1.0       -4.6        1.5   
1827  2002-01-02T00:00+00:00    13305   566.0   0.0       -3.0        2.1   
1828  2002-01-03T00:00+00:00    13305   684.0  -1.0       -3.7        2.1   
1829  2002-01-04T00:00+00:00    13305   668.0  -1.0      -11.7        1.0   
1830  2002-01-05T00:00+00:00    13305   235.0  -1.0      -12.5        1.0   
...                      ...      ...     ...   ...        ...        ...   
7806  2018-05-17T00:00+00:00    13305  1388.0  -1.0       14.4        1.0   
7807  2018-05-18T00:00+00:00    13305  1455.0  -1.0       15.9        1.0   
7808  2018-05-19T00:00+00:00    13305  2075.0  -1.0       15.9        2.1   
7809  2018-05-20T00:00+00:00    13305  1482.0  11.7       14.8        1.0   
7810  2018-05-21T00:00+00:00    13305  1771.0   1.4       16.1        1.0   

      p_mittel  tlmin  tlmax  
1826     972.4   -9.4    0.2  
1827     965.6   -9.8    3.8  
1828     971.8   -7.4    0.1  
1829     975.2  -16.8   -6.6  
1830     975.2  -16.7   -8.2  
...        ...    ...    ...  
7806     955.6    9.5   19.2  
7807     958.1   10.0   21.8  
7808     959.6    9.3   22.4  
7809     962.0    8.3   21.2  
7810     958.8   10.5   21.7  

[5959 rows x 9 columns]

This however drops all rows that contain a NaN in any place. clgo was not available until later (2002) so we loose a lot of data. We want to know at least when all temperature values were available)

In [3]:
df.dropna(subset=['tl_mittel', 'tlmax', 'tlmin']).head()

time  station  cglo_j   rr  tl_mittel  vv_mittel  \
59  1997-03-01T00:00+00:00    13305     NaN -1.0        2.4        1.0   
60  1997-03-02T00:00+00:00    13305     NaN -1.0        6.6        1.0   
61  1997-03-03T00:00+00:00    13305     NaN -1.0        6.2        0.6   
62  1997-03-04T00:00+00:00    13305     NaN  2.5        7.6        1.5   
63  1997-03-05T00:00+00:00    13305     NaN -1.0        6.4        0.6   

    p_mittel  tlmin  tlmax  
59     971.7   -1.9    6.6  
60     969.7   -0.5   13.6  
61     971.1   -1.2   13.6  
62     972.8    4.9   10.3  
63     968.0    3.3    9.4

In [4]:
df.dropna(subset=['tl_mittel', 'tlmax', 'tlmin']).tail()

time  station  cglo_j    rr  tl_mittel  vv_mittel  \
7807  2018-05-18T00:00+00:00    13305  1455.0  -1.0       15.9        1.0   
7808  2018-05-19T00:00+00:00    13305  2075.0  -1.0       15.9        2.1   
7809  2018-05-20T00:00+00:00    13305  1482.0  11.7       14.8        1.0   
7810  2018-05-21T00:00+00:00    13305  1771.0   1.4       16.1        1.0   
7811  2018-05-22T00:00+00:00    13305  1732.0   NaN       16.2        1.5   

      p_mittel  tlmin  tlmax  
7807     958.1   10.0   21.8  
7808     959.6    9.3   22.4  
7809     962.0    8.3   21.2  
7810     958.8   10.5   21.7  
7811     956.8    9.6   22.7

So from row 1826 until row 7810 seems to represent the time that the weather station was operational. We can simply slice the dataframe like we would a python list object.

In [11]:
df = df[59:7811].copy()

There are more elaborate ways to go about this, including not accessing the DataFrame using row indices, but using the time as index, but more on that later.

In [12]:
df


time  station  cglo_j    rr  tl_mittel  vv_mittel  \
118   1997-04-29T00:00+00:00    13305     NaN   0.1       11.1        0.6   
119   1997-04-30T00:00+00:00    13305     NaN  -1.0        9.7        1.5   
120   1997-05-01T00:00+00:00    13305     NaN  -1.0       11.0        1.5   
121   1997-05-02T00:00+00:00    13305     NaN  -1.0       12.6        1.5   
122   1997-05-03T00:00+00:00    13305     NaN  -1.0       14.9        1.5   
...                      ...      ...     ...   ...        ...        ...   
7806  2018-05-17T00:00+00:00    13305  1388.0  -1.0       14.4        1.0   
7807  2018-05-18T00:00+00:00    13305  1455.0  -1.0       15.9        1.0   
7808  2018-05-19T00:00+00:00    13305  2075.0  -1.0       15.9        2.1   
7809  2018-05-20T00:00+00:00    13305  1482.0  11.7       14.8        1.0   
7810  2018-05-21T00:00+00:00    13305  1771.0   1.4       16.1        1.0   

      p_mittel  tlmin  tlmax  tl_diff  
118      951.5    6.7   15.5      8.8  
119      958.0    4.5   14.8     10.3  
120      964.8    6.0   16.0     10.0  
121      966.8    4.9   20.2     15.3  
122      957.8    3.9   25.9     22.0  
...        ...    ...    ...      ...  
7806     955.6    9.5   19.2      9.7  
7807     958.1   10.0   21.8     11.8  
7808     959.6    9.3   22.4     13.1  
7809     962.0    8.3   21.2     12.9  
7810     958.8   10.5   21.7     11.2  

[7693 rows x 10 columns]

## Accessing columns

One can access separate columns (can be multiple) like we would with a dictionary:

In [13]:
df['tl_mittel']

118     11.1
119      9.7
120     11.0
121     12.6
122     14.9
        ... 
7806    14.4
7807    15.9
7808    15.9
7809    14.8
7810    16.1
Name: tl_mittel, Length: 7693, dtype: float64

In [14]:
df[['tl_mittel', 'tlmin', 'tlmax']]

tl_mittel  tlmin  tlmax
118        11.1    6.7   15.5
119         9.7    4.5   14.8
120        11.0    6.0   16.0
121        12.6    4.9   20.2
122        14.9    3.9   25.9
...         ...    ...    ...
7806       14.4    9.5   19.2
7807       15.9   10.0   21.8
7808       15.9    9.3   22.4
7809       14.8    8.3   21.2
7810       16.1   10.5   21.7

[7693 rows x 3 columns]

## Basic maths operations

Lets do some basic operations. Make a new column with the temperature differential tlmax - tmin.

In [15]:
df['tl_diff'] = df['tlmax'] - df['tlmin']
df.head()

time  station  cglo_j   rr  tl_mittel  vv_mittel  \
118  1997-04-29T00:00+00:00    13305     NaN  0.1       11.1        0.6   
119  1997-04-30T00:00+00:00    13305     NaN -1.0        9.7        1.5   
120  1997-05-01T00:00+00:00    13305     NaN -1.0       11.0        1.5   
121  1997-05-02T00:00+00:00    13305     NaN -1.0       12.6        1.5   
122  1997-05-03T00:00+00:00    13305     NaN -1.0       14.9        1.5   

     p_mittel  tlmin  tlmax  tl_diff  
118     951.5    6.7   15.5      8.8  
119     958.0    4.5   14.8     10.3  
120     964.8    6.0   16.0     10.0  
121     966.8    4.9   20.2     15.3  
122     957.8    3.9   25.9     22.0

**Exercise**: Create a new column named 'freezing' that contains True if the min temparature was below 0 and False otherwise.

In [17]:
df['freezing'] = df['tlmin'] < 0
df

time  station  cglo_j    rr  tl_mittel  vv_mittel  \
118   1997-04-29T00:00+00:00    13305     NaN   0.1       11.1        0.6   
119   1997-04-30T00:00+00:00    13305     NaN  -1.0        9.7        1.5   
120   1997-05-01T00:00+00:00    13305     NaN  -1.0       11.0        1.5   
121   1997-05-02T00:00+00:00    13305     NaN  -1.0       12.6        1.5   
122   1997-05-03T00:00+00:00    13305     NaN  -1.0       14.9        1.5   
...                      ...      ...     ...   ...        ...        ...   
7806  2018-05-17T00:00+00:00    13305  1388.0  -1.0       14.4        1.0   
7807  2018-05-18T00:00+00:00    13305  1455.0  -1.0       15.9        1.0   
7808  2018-05-19T00:00+00:00    13305  2075.0  -1.0       15.9        2.1   
7809  2018-05-20T00:00+00:00    13305  1482.0  11.7       14.8        1.0   
7810  2018-05-21T00:00+00:00    13305  1771.0   1.4       16.1        1.0   

      p_mittel  tlmin  tlmax  tl_diff  freezing  
118      951.5    6.7   15.5      8.8     False  
119      958.0    4.5   14.8     10.3     False  
120      964.8    6.0   16.0     10.0     False  
121      966.8    4.9   20.2     15.3     False  
122      957.8    3.9   25.9     22.0     False  
...        ...    ...    ...      ...       ...  
7806     955.6    9.5   19.2      9.7     False  
7807     958.1   10.0   21.8     11.8     False  
7808     959.6    9.3   22.4     13.1     False  
7809     962.0    8.3   21.2     12.9     False  
7810     958.8   10.5   21.7     11.2     False  

[7693 rows x 11 columns]

**Exercise**: Instead of True and False this freezing column should contain 1 and 0 (1 if True, 0 if False)

In [23]:
df['freezing'] = (df['tlmin'] < 0).astype(int)
df

time  station  cglo_j    rr  tl_mittel  vv_mittel  \
118   1997-04-29T00:00+00:00    13305     NaN   0.1       11.1        0.6   
119   1997-04-30T00:00+00:00    13305     NaN  -1.0        9.7        1.5   
120   1997-05-01T00:00+00:00    13305     NaN  -1.0       11.0        1.5   
121   1997-05-02T00:00+00:00    13305     NaN  -1.0       12.6        1.5   
122   1997-05-03T00:00+00:00    13305     NaN  -1.0       14.9        1.5   
...                      ...      ...     ...   ...        ...        ...   
7806  2018-05-17T00:00+00:00    13305  1388.0  -1.0       14.4        1.0   
7807  2018-05-18T00:00+00:00    13305  1455.0  -1.0       15.9        1.0   
7808  2018-05-19T00:00+00:00    13305  2075.0  -1.0       15.9        2.1   
7809  2018-05-20T00:00+00:00    13305  1482.0  11.7       14.8        1.0   
7810  2018-05-21T00:00+00:00    13305  1771.0   1.4       16.1        1.0   

      p_mittel  tlmin  tlmax  tl_diff  freezing  
118      951.5    6.7   15.5      8.8         0  
119      958.0    4.5   14.8     10.3         0  
120      964.8    6.0   16.0     10.0         0  
121      966.8    4.9   20.2     15.3         0  
122      957.8    3.9   25.9     22.0         0  
...        ...    ...    ...      ...       ...  
7806     955.6    9.5   19.2      9.7         0  
7807     958.1   10.0   21.8     11.8         0  
7808     959.6    9.3   22.4     13.1         0  
7809     962.0    8.3   21.2     12.9         0  
7810     958.8   10.5   21.7     11.2         0  

[7693 rows x 11 columns]

0

## Basic stats

Get some basic statistics on the data using describe().

In [24]:
df.describe()

station       cglo_j           rr    tl_mittel    vv_mittel  \
count   7693.0  5959.000000  7693.000000  7693.000000  7693.000000   
mean   13305.0  1128.234435     1.684752     9.938711     1.081698   
std        0.0   733.033308     6.081223     8.210142     0.508782   
min    13305.0    47.000000    -1.000000   -14.200000     0.000000   
25%    13305.0   486.000000    -1.000000     3.200000     0.600000   
50%    13305.0   998.000000    -1.000000    10.500000     1.000000   
75%    13305.0  1690.000000     1.200000    16.800000     1.500000   
max    13305.0  3085.000000    77.400000    27.900000     5.900000   

          p_mittel        tlmin        tlmax      tl_diff     freezing  
count  7693.000000  7693.000000  7693.000000  7693.000000  7693.000000  
mean    957.614754     4.854959    14.985168    10.130209     0.278565  
std       7.230782     7.390349     9.525686     4.667688     0.448322  
min     919.800000   -20.000000    -9.300000     1.000000     0.000000  
25%     953.400000    -0.600000     6.700000     6.500000     0.000000  
50%     957.900000     5.100000    15.800000     9.700000     0.000000  
75%     962.200000    11.100000    22.700000    13.500000     1.000000  
max     980.400000    21.700000    37.200000    24.600000     1.000000

## Conditional slicing, finding and counting occurrences

When was the coldest day in Kapfenberg? We can see above that tlmin had a lowest value of -20, but when?
We can use conditions that evaluate to true or false (like the freezing one above) as indexers.

In [25]:
df['tlmin'] == -20.0

118     False
119     False
120     False
121     False
122     False
        ...  
7806    False
7807    False
7808    False
7809    False
7810    False
Name: tlmin, Length: 7693, dtype: bool

In [26]:
df[df['tlmin'] == -20.0]


time  station  cglo_j   rr  tl_mittel  vv_mittel  \
2202  2003-01-12T00:00+00:00    13305   691.0 -1.0      -14.2        1.0   

      p_mittel  tlmin  tlmax  tl_diff  freezing  
2202     970.8  -20.0   -8.4     11.6         1

We can use this to select whole ranges of data where some condition applies. E.g. select all data where it was freezing.

In [27]:
df['freezing'] = (df['tlmin'] < 0).astype(int)
df[df['freezing'] == 1]

time  station  cglo_j   rr  tl_mittel  vv_mittel  \
286   1997-10-14T00:00+00:00    13305     NaN  0.6        4.9        1.5   
297   1997-10-25T00:00+00:00    13305     NaN -1.0        2.4        2.1   
299   1997-10-27T00:00+00:00    13305     NaN  1.8       -0.4        1.0   
300   1997-10-28T00:00+00:00    13305     NaN -1.0       -1.8        2.6   
301   1997-10-29T00:00+00:00    13305     NaN -1.0       -2.7        1.0   
...                      ...      ...     ...  ...        ...        ...   
7751  2018-03-23T00:00+00:00    13305  1129.0  0.0        3.4        0.6   
7752  2018-03-24T00:00+00:00    13305   953.0 -1.0        4.4        1.0   
7753  2018-03-25T00:00+00:00    13305  1215.0 -1.0        2.9        1.5   
7756  2018-03-28T00:00+00:00    13305  1033.0  0.2        3.8        1.0   
7766  2018-04-07T00:00+00:00    13305  1887.0 -1.0        8.8        1.5   

      p_mittel  tlmin  tlmax  tl_diff  freezing  
286      953.5   -0.5   10.2     10.7         1  
297      962.2   -3.5    8.3     11.8         1  
299      965.4   -3.9    3.1      7.0         1  
300      972.8   -6.5    3.0      9.5         1  
301      976.7   -8.8    3.5     12.3         1  
...        ...    ...    ...      ...       ...  
7751     945.8   -1.0    7.8      8.8         1  
7752     945.6   -0.3    9.1      9.4         1  
7753     949.8   -1.3    7.0      8.3         1  
7756     950.0   -2.3    9.9     12.2         1  
7766     960.7   -0.2   17.7     17.9         1  

[2143 rows x 11 columns]

We can use value_counts to see what value occurs how often in the dataframe.

In [30]:
df['freezing'].value_counts()

freezing
0    5550
1    2143
Name: count, dtype: int64

**Exercise**: How many days were the temperature was always freezing (look at tlmax) and what percentage of the time does this represent.

In [34]:
df['allDayFreezing'] = (df['tlmax'] < 0).astype(int)
counts = df['allDayFreezing'].value_counts()
counts

allDayFreezing
0    7325
1     368
Name: count, dtype: int64

In [36]:
counts[1] / (counts[0] + counts[1]) *100

4.783569478746913

## Time indexed DataFrames

In [37]:
df

time  station  cglo_j    rr  tl_mittel  vv_mittel  \
118   1997-04-29T00:00+00:00    13305     NaN   0.1       11.1        0.6   
119   1997-04-30T00:00+00:00    13305     NaN  -1.0        9.7        1.5   
120   1997-05-01T00:00+00:00    13305     NaN  -1.0       11.0        1.5   
121   1997-05-02T00:00+00:00    13305     NaN  -1.0       12.6        1.5   
122   1997-05-03T00:00+00:00    13305     NaN  -1.0       14.9        1.5   
...                      ...      ...     ...   ...        ...        ...   
7806  2018-05-17T00:00+00:00    13305  1388.0  -1.0       14.4        1.0   
7807  2018-05-18T00:00+00:00    13305  1455.0  -1.0       15.9        1.0   
7808  2018-05-19T00:00+00:00    13305  2075.0  -1.0       15.9        2.1   
7809  2018-05-20T00:00+00:00    13305  1482.0  11.7       14.8        1.0   
7810  2018-05-21T00:00+00:00    13305  1771.0   1.4       16.1        1.0   

      p_mittel  tlmin  tlmax  tl_diff  freezing  allDayFreezing  
118      951.5    6.7   15.5      8.8         0               0  
119      958.0    4.5   14.8     10.3         0               0  
120      964.8    6.0   16.0     10.0         0               0  
121      966.8    4.9   20.2     15.3         0               0  
122      957.8    3.9   25.9     22.0         0               0  
...        ...    ...    ...      ...       ...             ...  
7806     955.6    9.5   19.2      9.7         0               0  
7807     958.1   10.0   21.8     11.8         0               0  
7808     959.6    9.3   22.4     13.1         0               0  
7809     962.0    8.3   21.2     12.9         0               0  
7810     958.8   10.5   21.7     11.2         0               0  

[7693 rows x 12 columns]

Pandas supports datetime indices

In [40]:
df['time'] = pandas.to_datetime(df['time'], utc=True)
# df
df = df.set_index('time', drop=True)

In [41]:
df.head()


station  cglo_j   rr  tl_mittel  vv_mittel  \
time                                                                    
1997-04-29 00:00:00+00:00    13305     NaN  0.1       11.1        0.6   
1997-04-30 00:00:00+00:00    13305     NaN -1.0        9.7        1.5   
1997-05-01 00:00:00+00:00    13305     NaN -1.0       11.0        1.5   
1997-05-02 00:00:00+00:00    13305     NaN -1.0       12.6        1.5   
1997-05-03 00:00:00+00:00    13305     NaN -1.0       14.9        1.5   

                           p_mittel  tlmin  tlmax  tl_diff  freezing  \
time                                                                   
1997-04-29 00:00:00+00:00     951.5    6.7   15.5      8.8         0   
1997-04-30 00:00:00+00:00     958.0    4.5   14.8     10.3         0   
1997-05-01 00:00:00+00:00     964.8    6.0   16.0     10.0         0   
1997-05-02 00:00:00+00:00     966.8    4.9   20.2     15.3         0   
1997-05-03 00:00:00+00:00     957.8    3.9   25.9     22.0         0   

                           allDayFreezing  
time                                       
1997-04-29 00:00:00+00:00               0  
1997-04-30 00:00:00+00:00               0  
1997-05-01 00:00:00+00:00               0  
1997-05-02 00:00:00+00:00               0  
1997-05-03 00:00:00+00:00               0

Now we can index rows based on times.

In [42]:
df['2011-03-01 00:00': '2011-03-02 00:00']

station  cglo_j   rr  tl_mittel  vv_mittel  \
time                                                                    
2011-03-01 00:00:00+00:00    13305   194.0 -1.0       -0.4        1.5   
2011-03-02 00:00:00+00:00    13305   501.0 -1.0       -2.6        1.5   

                           p_mittel  tlmin  tlmax  tl_diff  freezing  \
time                                                                   
2011-03-01 00:00:00+00:00     969.1   -4.3    3.6      7.9         1   
2011-03-02 00:00:00+00:00     969.5   -7.2    2.0      9.2         1   

                           allDayFreezing  
time                                       
2011-03-01 00:00:00+00:00               0  
2011-03-02 00:00:00+00:00               0

We can now resample the dataframe to some other resolution. Resampling to a lower frequency is called downsampling, to a higher frequency this is called upsampling.

When resampling, one has to specify a frequency and a method.

For example yearly avarages:


In [51]:
df.resample('1YS').mean() 

station       cglo_j        rr  tl_mittel  \
time                                                                   
1997-01-01 00:00:00+00:00  13305.0          NaN  2.032389  12.071255   
1998-01-01 00:00:00+00:00  13305.0          NaN  1.438904   9.493699   
1999-01-01 00:00:00+00:00  13305.0          NaN  1.624384   9.472603   
2000-01-01 00:00:00+00:00  13305.0          NaN  1.778415  10.279508   
2001-01-01 00:00:00+00:00  13305.0          NaN  0.851507   9.569589   
2002-01-01 00:00:00+00:00  13305.0  1167.607735  1.838082  10.171781   
2003-01-01 00:00:00+00:00  13305.0  1215.260989  1.054795   9.716986   
2004-01-01 00:00:00+00:00  13305.0  1083.517808  1.627869   9.137705   
2005-01-01 00:00:00+00:00  13305.0  1179.101928  2.273425   8.987123   
2006-01-01 00:00:00+00:00  13305.0  1191.417582  1.355890   9.509041   
2007-01-01 00:00:00+00:00  13305.0  1186.676796  1.912603  10.420822   
2008-01-01 00:00:00+00:00  13305.0  1156.145604  1.227322  10.375410   
2009-01-01 00:00:00+00:00  13305.0  1216.916898  2.200274   9.984658   
2010-01-01 00:00:00+00:00  13305.0  1046.597260  1.884110   8.976712   
2011-01-01 00:00:00+00:00  13305.0  1083.458333  1.427945  10.151233   
2012-01-01 00:00:00+00:00  13305.0  1147.753425  2.221585  10.037158   
2013-01-01 00:00:00+00:00  13305.0  1119.401099  2.204658   9.908219   
2014-01-01 00:00:00+00:00  13305.0  1027.317808  1.958356  10.963288   
2015-01-01 00:00:00+00:00  13305.0  1083.868132  1.344384  10.635616   
2016-01-01 00:00:00+00:00  13305.0  1087.279452  1.798361  10.618579   
2017-01-01 00:00:00+00:00  13305.0  1108.347945  1.578356  10.300822   
2018-01-01 00:00:00+00:00  13305.0  1006.219858  1.317021   6.362411   

                           vv_mittel    p_mittel     tlmin      tlmax  \
time                                                                    
1997-01-01 00:00:00+00:00   1.327126  957.073279  7.094737  17.012551   
1998-01-01 00:00:00+00:00   1.167945  957.485205  4.482466  14.470137   
1999-01-01 00:00:00+00:00   1.095890  956.593973  4.757534  14.155890   
2000-01-01 00:00:00+00:00   1.066667  956.969945  5.255464  15.263661   
2001-01-01 00:00:00+00:00   1.067123  956.835342  4.418904  14.687945   
2002-01-01 00:00:00+00:00   1.174795  957.592877  5.231233  15.073151   
2003-01-01 00:00:00+00:00   1.172877  959.304658  4.058082  15.339178   
2004-01-01 00:00:00+00:00   1.126503  958.069399  4.332514  13.911202   
2005-01-01 00:00:00+00:00   1.205753  959.006575  4.060000  13.883562   
2006-01-01 00:00:00+00:00   1.218904  959.675068  4.452877  14.530411   
2007-01-01 00:00:00+00:00   1.096986  958.124658  5.119178  15.675616   
2008-01-01 00:00:00+00:00   1.058197  957.318852  5.303552  15.404918   
2009-01-01 00:00:00+00:00   1.003562  956.141096  5.116438  14.810411   
2010-01-01 00:00:00+00:00   0.996438  954.512603  4.271781  13.651233   
2011-01-01 00:00:00+00:00   1.002466  959.361918  4.749315  15.512329   
2012-01-01 00:00:00+00:00   1.061749  957.536612  4.793989  15.241803   
2013-01-01 00:00:00+00:00   0.987397  956.576164  5.135068  14.644110   
2014-01-01 00:00:00+00:00   1.006301  956.577808  6.167671  15.717808   
2015-01-01 00:00:00+00:00   1.001370  959.734521  5.241918  15.991781   
2016-01-01 00:00:00+00:00   0.918579  958.090984  5.268852  15.924863   
2017-01-01 00:00:00+00:00   0.971781  958.513425  4.631781  15.935068   
2018-01-01 00:00:00+00:00   1.255319  954.094326  1.575177  11.112766   

                             tl_diff  freezing  allDayFreezing  
time                                                            
1997-01-01 00:00:00+00:00   9.917814  0.129555        0.016194  
1998-01-01 00:00:00+00:00   9.987671  0.301370        0.043836  
1999-01-01 00:00:00+00:00   9.398356  0.304110        0.052055  
2000-01-01 00:00:00+00:00  10.008197  0.232240        0.051913  
2001-01-01 00:00:00+00:00  10.269041  0.298630        0.060274  
2002-01-01 00:00:00+00:00   9.841918  0.249315        0.052055  
2003-01-01 00:00:00+00:00

We can also resample to a higher frequency than the original data.

For example upsampling to hourly frequency while using linear interpolation.

In [52]:
small_df = df['2011-03-01 00:00': '2011-03-07 00:00'].copy() 
small_df.resample('1h').interpolate()


station       cglo_j   rr  tl_mittel  vv_mittel  \
time                                                                         
2011-03-01 00:00:00+00:00  13305.0   194.000000 -1.0  -0.400000   1.500000   
2011-03-01 01:00:00+00:00  13305.0   206.791667 -1.0  -0.491667   1.500000   
2011-03-01 02:00:00+00:00  13305.0   219.583333 -1.0  -0.583333   1.500000   
2011-03-01 03:00:00+00:00  13305.0   232.375000 -1.0  -0.675000   1.500000   
2011-03-01 04:00:00+00:00  13305.0   245.166667 -1.0  -0.766667   1.500000   
...                            ...          ...  ...        ...        ...   
2011-03-06 20:00:00+00:00  13305.0  1449.500000 -1.0   0.450000   1.416667   
2011-03-06 21:00:00+00:00  13305.0  1456.375000 -1.0   0.387500   1.437500   
2011-03-06 22:00:00+00:00  13305.0  1463.250000 -1.0   0.325000   1.458333   
2011-03-06 23:00:00+00:00  13305.0  1470.125000 -1.0   0.262500   1.479167   
2011-03-07 00:00:00+00:00  13305.0  1477.000000 -1.0   0.200000   1.500000   

                             p_mittel     tlmin     tlmax   tl_diff  freezing  \
time                                                                            
2011-03-01 00:00:00+00:00  969.100000 -4.300000  3.600000  7.900000       1.0   
2011-03-01 01:00:00+00:00  969.116667 -4.420833  3.533333  7.954167       1.0   
2011-03-01 02:00:00+00:00  969.133333 -4.541667  3.466667  8.008333       1.0   
2011-03-01 03:00:00+00:00  969.150000 -4.662500  3.400000  8.062500       1.0   
2011-03-01 04:00:00+00:00  969.166667 -4.783333  3.333333  8.116667       1.0   
...                               ...       ...       ...       ...       ...   
2011-03-06 20:00:00+00:00  970.633333 -4.083333  4.900000  8.983333       1.0   
2011-03-06 21:00:00+00:00  971.125000 -4.112500  4.800000  8.912500       1.0   
2011-03-06 22:00:00+00:00  971.616667 -4.141667  4.700000  8.841667       1.0   
2011-03-06 23:00:00+00:00  972.108333 -4.170833  4.600000  8.770833       1.0   
2011-03-07 00:00:00+00:00  972.600000 -4.200000  4.500000  8.700000       1.0   

                           allDayFreezing  
time                                       
2011-03-01 00:00:00+00:00             0.0  
2011-03-01 01:00:00+00:00             0.0  
2011-03-01 02:00:00+00:00             0.0  
2011-03-01 03:00:00+00:00             0.0  
2011-03-01 04:00:00+00:00             0.0  
...                                   ...  
2011-03-06 20:00:00+00:00             0.0  
2011-03-06 21:00:00+00:00             0.0  
2011-03-06 22:00:00+00:00             0.0  
2011-03-06 23:00:00+00:00             0.0  
2011-03-07 00:00:00+00:00             0.0  

[145 rows x 11 columns]

**Exercise**: Ressample the 'freezing' column to yearly frequency providing not the mean (as in the examples above) but the sum within each year.

In [55]:
df['freezing'].resample('1YS').sum() 

time
1997-01-01 00:00:00+00:00     32
1998-01-01 00:00:00+00:00    110
1999-01-01 00:00:00+00:00    111
2000-01-01 00:00:00+00:00     85
2001-01-01 00:00:00+00:00    109
2002-01-01 00:00:00+00:00     91
2003-01-01 00:00:00+00:00    126
2004-01-01 00:00:00+00:00    120
2005-01-01 00:00:00+00:00    127
2006-01-01 00:00:00+00:00    108
2007-01-01 00:00:00+00:00     87
2008-01-01 00:00:00+00:00     89
2009-01-01 00:00:00+00:00     93
2010-01-01 00:00:00+00:00    115
2011-01-01 00:00:00+00:00    119
2012-01-01 00:00:00+00:00     94
2013-01-01 00:00:00+00:00     97
2014-01-01 00:00:00+00:00     66
2015-01-01 00:00:00+00:00    103
2016-01-01 00:00:00+00:00     91
2017-01-01 00:00:00+00:00    106
2018-01-01 00:00:00+00:00     64
Freq: YS-JAN, Name: freezing, dtype: int64